cmd/agent/schedule_agent_service.go

```diff
@@ -0,0 +1,337 @@
+	for id, item := range a.schedule {
+		if _, exists := actionMap[id]; !exists {
+			item.cancel()
+			delete(a.schedule, id)
+			a.logger.Warn("Action Cancelled", zap.String("action_id", id))
+		}
+
+	}
+
+	// Checking the entire new schedule to schedule or reschedule actions
+	for _, action := range newSchedule {
+		var scheduledFunc func(actions.ScheduledAction)
+		var cronFunc func(actions.CronAction)
+
+		if _, exists := a.schedule[action.GetID()]; !exists { // Schedule new actions
+			scheduledFunc = a.scheduleNewScheduledAction
+			cronFunc = a.scheduleNewCronAction
+		} else { // Reschedule actions
+			scheduledFunc = a.rescheduleScheduledAction
+			cronFunc = a.rescheduleCronAction
+		}
+
+		// managing actions based on type
+		switch action.(type) {
+		case actions.ScheduledAction:
+			scheduledFunc(action.(actions.ScheduledAction))
+		case actions.CronAction:
+			cronFunc(action.(actions.CronAction))
+		default:
+			a.logger.Error("Unknown action type", zap.String("action_id", action.GetID()))
+		}
+	}
+
+}
+
+// fetchScheduledActions goes to the API and retrieves the updated list of scheduled actions on the DB
+//
+// Parameters:
+//
+// Returns:
+//   - A list of the actions.ScheduledAction retrieved from the API
+//   - An error if the API querying fails
+func (a *ScheduleAgentService) fetchScheduledActions() (*[]actions.Action, error) {
+	var b []byte
+	// Prepare API request
+	request, err := http.NewRequest(http.MethodGet, a.cfg.APIURL+API_SCHEDULE_ACTIONS_PATH, bytes.NewBuffer(b))
+	if err != nil {
+		return nil, err
+	}
+
+	// Adding query parameter for the status. Only "Pending" or enabled actions are retrieved
+	q := request.URL.Query()
+	q.Add("status", "Pending")
+	q.Add("enabled", "true")
+	request.URL.RawQuery = q.Encode()
+
+	// Performing API request
+	response, err := a.client.Do(request)
+	if err != nil {
+		return nil, err
+	}
+
+	// Reading response body
+	body, err := io.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	// Struct for Unmarshalling results
+	var result struct {
+		Count   int               `json:"count"`
+		Actions []json.RawMessage `json:"actions"`
+	}
+
+	// Unmarshalling response
+	err = json.Unmarshal(body, &result)
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshalling Actions by type
+	resultActions, err := actions.DecodeActions(result.Actions)
+	if err != nil {
+		return nil, err
+	}
+
+	a.logger.Debug("Fetched scheduled actions", zap.Int("actions_num", len(*resultActions)))
+
+	return resultActions, nil
+}
+
+// ReScheduleActions maintains an infinite loop for rescheduling the actions
+//
+// Parameters:
+//
+// Returns:
+func (a *ScheduleAgentService) ReScheduleActions() {
+	// Ticker for performing the polling loop
+	ticker := time.NewTicker(time.Duration(a.cfg.PollingInterval) * time.Second)
+	defer ticker.Stop()
+
+	for {
+		a.logger.Debug("Polling Schedule from DB")
+		if actions, err := a.fetchScheduledActions(); err != nil {
```

Comment:
I guess would be better to avoid package collides

Reply:
replaced `actions` by `fetchedActions`